### ISSUE TITLE
Notifications: Native slack target

### ISSUE THREAD

**Issue by stirby** [coder/coder#14740]
Our notifications currently support SMTP and Webhook as delivery targets. We want to add native support for Slack workspaces as many customers use this as their internal messaging platform. 

The native Slack integration should allow administrators to direct notifications towards a Slack workspace. Users should receive notifications via a Slack app in that workspace, not a channel. 

This product work should include server flags to set the Slack target and a new option in the deployment notifications settings UI for the delivery target. This should be paired with adding docs explaining how to connect our notifications with Slack.

> Note: We'll only be adding MSFT Teams and Slack as natively supported delivery methods.

**Comment by mtojek** [comment]
I did the reconnaissance, and there won't be any shortcuts like we had for MS Teams. There are 2 options:

1. Slack Workflows:
* Slack indeed supports [Workflows](https://api.slack.com/automation/create), but they are limited - incoming webhooks support only root-level keys. Looks like we need to add a parameter to customize/transform the payload?
* I have concerns regarding their pricing model: [Standard and premium workflow usage](https://slack.com/help/articles/15363357403411-Pricing-guide-for-Slack-Workflow-Builder). Isn't too expensive for an average usage?
* Other than that, it should be fairly simple using the recommended framework: [deno-slack-sdk](https://github.com/slackapi/deno-slack-sdk) and [custom functions](https://api.slack.com/automation/functions/custom)

2. Build the old-school self-hosted Slack app:
* Use [Bolt for JS](https://tools.slack.dev/bolt-js/getting-started/) to run the bot app
* It just interacts with Slack API and have its own webserver so we don't have to modify coder/coder

Thoughts @dannykopping?

**Comment by dannykopping** [comment]
> Other than that, it should be fairly simple using the recommended framework: [deno-slack-sdk](https://github.com/slackapi/deno-slack-sdk) and [custom functions](https://api.slack.com/automation/functions/custom)

🤔 is seems like you are proposing we implement this in the frontend? These notifications would be coming from `coderd`.

Did you look into [Slack's webhooks](https://api.slack.com/messaging/webhooks) + Slack app? This is how I've built Slack integrations in the past, and it seems pretty simple.

**Comment by mtojek** [comment]
> 🤔 is seems like you are proposing we implement this in the frontend? These notifications would be coming from coderd.

Frankly speaking, I would vote for option 2 and don't touch `coderd` at all, but I won't play favorites here.

> Did you look into [Slack's webhooks](https://api.slack.com/messaging/webhooks) + Slack app?

I looked at this page, and it seems that you can configure incoming webhooks to post to single channel. I'm afraid we can't use this method as we need to post messages to specific users.

**Comment by dannykopping** [comment]
> Frankly speaking, I would vote for option 2 and don't touch coderd at all, but I won't play favorites here.

I don't understand, though; why would Slack be different from the other notification methods? Maybe I'm missing some context here. From a user perspective, I wouldn't want to have to run another piece of software to enable Slack notifications; I would expect it to "just work" and use a webhook or API client.

> I looked at this page, and it seems that you can configure incoming webhooks to post to single channel. I'm afraid we can't use this method as we need to post messages to specific users.

Yeah it appears we'd need access to the API for that. I think this could be fine, since Slack API keys can be narrowly scoped.

https://api.slack.com/methods/chat.postMessage
https://api.slack.com/methods/users.lookupByEmail
There's a Go client lib which can interact with the API: https://pkg.go.dev/github.com/slack-go/slack#Client.PostMessage.

Otherwise we could add a field to the user table which specifies the user's Slack handle, but that seems a step too far.

**Comment by stirby** [comment]
Firstly, I agree with Danny that the admin experience would require hitting the slack API directly. However, I've already received negative internal feedback on making this a native integration for the reason Marcin mentioned:

> Having a webhook sidecar for Slack would prevent us from polluting coder/coder with 3rd party integrations with other platforms like Teams, Discord, Mattermost, etc.

I suggest we write documentation to create a Slack-hosted workflow for our notifications, as the companies requesting this integration have far more mature Slack workspaces than ours. I expect them to be comfortable with this tooling. If we see great resistance, we can revisit native integration.

I can reach out to customers and gauge familiarity with hosted workflows.

**Comment by dannykopping** [comment]
From my side, all I want is user simplicity. Alertmanager is a very widely deployed service and [just hits the API directly](https://prometheus.io/docs/alerting/latest/configuration/#slack_config); super simple.

If setting up a workflow is just as simple, then I'm fine with this approach.

**Comment by mtojek** [comment]
For the self-hosted option, admins can run the following script:

```js
const { App, LogLevel, ExpressReceiver } = require("@slack/bolt");
const bodyParser = require("body-parser");

const port = process.env.PORT || 6000;

// Create a Bolt Receiver
const receiver = new ExpressReceiver({
  signingSecret: process.env.SLACK_SIGNING_SECRET,
});
receiver.router.use(bodyParser.json());

// Create the Bolt App, using the receiver
const app = new App({
  token: process.env.SLACK_BOT_TOKEN,
  logLevel: LogLevel.DEBUG,
  receiver,
});

receiver.router.post("/webhook", async (req, res) => {
  try {
    if (!req.body) {
      return res.status(400).send("Error: request body is missing");
    }

    const { title, body } = req.body;
    if (!title || !body) {
      return res.status(400).send('Error: missing fields: "title", or "body"');
    }

    const payload = req.body.payload;
    if (!payload) {
      return res.status(400).send('Error: missing "payload" field');
    }

    const { user_email, actions } = payload;
    if (!user_email || !actions) {
      return res
        .status(400)
        .send('Error: missing fields: "user_email", "actions"');
    }

    // Get the user ID using Slack API
    const userByEmail = await app.client.users.lookupByEmail({
      email: user_email,
    });

    const slackMessage = {
      channel: userByEmail.user.id,
      text: body,
      blocks: [
        {
          type: "header",
          text: { type: "plain_text", text: title },
        },
        {
          type: "section",
          text: { type: "mrkdwn", text: body },
        },
      ],
    };

    // Add action buttons if they exist
    if (actions && actions.length > 0) {
      slackMessage.blocks.push({
        type: "actions",
        elements: actions.map((action) => ({
          type: "button",
          text: { type: "plain_text", text: action.label },
          url: action.url,
        })),
      });
    }

    // Post message to the user on Slack
    await app.client.chat.postMessage(slackMessage);

    res.status(204).send();
  } catch (error) {
    console.error("Error sending message:", error);
    res.status(500).send();
  }
});

// Acknowledge clicks on link_button, otherwise Slack UI
// complains about missing events.
app.action("button_click", async ({ body, ack, say }) => {
  await ack(); // no specific action needed
});

// Start the Bolt app
(async () => {
  await app.start(port);
  console.log("⚡️ Coder Slack bot is running!");
})();
```

It results in: 

<img width="433" alt="Screenshot 2024-09-26 at 14 51 34" src="https://github.com/user-attachments/assets/3471b4c1-f2f8-45cf-9d6f-130c434c8f6d">

I will proceed with a docs write-up on how to enable the Slack bot.

**Comment by dannykopping** [comment]
This ticket is for the native Slack target; we're meeting on Monday to decide if we want to go that route or if @mtojek's Bolt app mentioned above suffices for now.

**Comment by mtojek** [comment]
TL;DR agreed to build a native integration for Slack.

We can talk about configuration/UI details later.

**Comment by defelmnq** [comment]
A few points I'd like to validate based on the current status.

## Application type

### Current situation

Based on [this part of the doc](https://coder.com/docs/admin/monitoring/notifications/slack), Coder can send notifications through Slack to users with a webserver started on their side. This solution requires some work and maintaining some pieces of code from the users, which we'd like to avoid.

### Proposals

Two possibilities seem to make sense based on our requirements. 

- using [incoming webhooks](https://codercom.slack.com/marketplace/A0F7XDUAZ-incoming-webhooks). This solution would be the simplest one but comes with three main points : 
  - It is defined as deprecated by the Slack team and is not recommended.
  - It does not allow us to ping either users or channels as we want, each webhook being connected to a given channel.
  - We do not have any access to the Slack information, and for example can not search for users by email etc...
- creating a [Slack app](https://api.slack.com/docs) - specifically I would recommend a `Non-workflow app`
  - Give us more control over the Slack workspace, with for example users and search
  - Can lead to companies refusing to use it as it exposes more of their information.
  - Requires more work on our side for the Application Connection between Slack and Coder.

Based on that, I would vote for the Slack App (second option) but think of a step by step solution to avoid huge unreviewable PR.

## Notifications

Another question is on how we want these notifications to be sent. 
The way [our notifications system](https://github.com/coder/coder/tree/main/coderd/notifications) currently works is by sending information to users, one by one individually. 

Slack works a bit differently and it can make sense to think about if we want to send these notifications to each user or, for example, a channel (or set of channels) defined by the users.

### Proposals

As said, two solutions here : 

- using the already existing logic and send notifications to each user / admin individually. (compatible with Slack App)
  - This solution is simple to implement as it relies on the currently existing logic.
  - It would require us to use the `External Authentication` tab - more specifically Slack Notify part and see if, using these information, we can send messages to users (with the Slack App above.)
  - It means that if a user / an admin has not connected its Slack profile, they can miss notifications.
- using channels to forward notifications. (Compatible with Slack App and Incoming Webhook)
  - This solution requires a new logic in the notifier to send notifications. We do not rely on a user anymore.
  - It would avoid users potentially missing notifications - also will allow a better communication and visibility for users in their Slack workspaces (threads, ping, messages...) 

I would say that the solution to choose depends on how much work we want to push on it - at least for the first iteration. If we go for the first solution here, we just need to validate that we are good with having some notifications potentially lost for users without the Slack notify connected.

More than keen to have some ideas and opinions from @dannykopping @mtojek @johnstcn @bpmct

**Comment by stirby** [comment]
Hey @defelmnq. Thank you for the investigation and this write-up.


> using the already existing logic and send notifications to each user / admin individually. 

I believe we should maintain this. Our notification system has been designed around delivery to a single user, so I'd advocate for the Slack app approach. We should consider methods of enabling batch delivery (delivery to all users in a group/organization, all users on a template) for custom notifications, but I don't think Slack should be the motivation to refactor this.

 
> It means that if a user / an admin has not connected its Slack profile, they can miss notifications.

Right now, delivery method is configurable by the administrator. Admins are responsible for setting up integration points and then assigning each notification one of the configured delivery methods (Webhook and SMTP right now). This is an anti-pattern since administrators should not be responsible for guessing how a user wants to receive alerts. The selection of delivery method (Webhook, SMTP, Slack, Teams) will instead be moved to the user-level config, as notifications are personal. 

From my perspective, if a user wants Slack notifications, some integration steps at the account level can be assumed.


> I would say that the solution to choose depends on how much work we want to push on it - at least for the first iteration.

We have to resolve a few things before committing to this.
- Implement user-level delivery configuration 
- Collect customer feedback on the existing integration method
- Estimate lift on implementing the slack app

**Comment by dannykopping** [comment]
I agree that the Slack App is the better approach to pursue.

@defelmnq you mention this:

> Can lead to companies refusing to use it as it exposes more of their information.

Can you expand upon that please?

> It would require us to use the `External Authentication` tab - more specifically Slack Notify part and see if, using these information, we can send messages to users (with the Slack App above.)

I don't think that the user needs to authenticate with Slack in order to link their accounts, though.
If we have the Slack App I believe we can query the API to resolve a Slack username for a given email address, no?

@stirby 

> Right now, delivery method is configurable by the administrator. Admins are responsible for setting up integration points and then assigning each notification one of the configured delivery methods (Webhook and SMTP right now). This is an anti-pattern since administrators should not be responsible for guessing how a user wants to receive alerts. The selection of delivery method (Webhook, SMTP, Slack, Teams) will instead be moved to the user-level config, as notifications are personal.

If this is the case then we need to pause this work and write up an RFC for user-level configurability. If we do go for this, then yeah we could have users link their Slack accounts instead of looking them up by email, which would be cleaner.

**Comment by defelmnq** [comment]
Thanks for all the feedback here, indeed what I had in mind is, overall (for the two first points), that I don't know if we want to rely only on the email address users use to register to Coder. 

If for example they use alias-based email (`user+coder@gmail.com` & `user+slack@gmail.com`) or simply different email address, we'll potentially fail to lookup for the user.

Also trying to think more generically, we'll potentially not be able to use the same methods in other platforms later (Teams, ...) and having a way for users to fill their information for each platform can improve this situation.

**Comment by dannykopping** [comment]
Yeah that's true, good points 👍
